### PR TITLE
[BitSail][Connector] Support Assert sink connector

### LINKED ISSUE #153: [BitSail][Connector] Support Assert sink connector
## Is your feature request related to a problem? Please describe

Currently we support print sink and we use print in some tests, but print is not sufficient enough to verify the test result.
So we need an `AssertSink` to verify the test result more accurately.

### PR DESCRIPTION
**Pull request by liuxiaocs7**
Signed-off-by:

## Pre-Checklist

Note: Please complete **_ALL_** items in the following checklist.

- [x] I have read through the [CONTRIBUTING.md](https://github.com/bytedance/bitsail/blob/master/docs/contributing.md) documentation.
- [x] My code has the necessary comments and documentation (if needed).
- [x] I have added relevant tests.

## Purpose

Some description about what this PR wants to do.

## Approaches

Currently we support print sink and we use print in some tests, but print is not sufficient enough to verify the test result.
So we implement an AssertSink to verify the test result more accurately.

## Related Issues

Close #153

## New Behavior (screenshots if needed)

N/A

### CODE REVIEW

File: bitsail-connectors/connector-assert/src/main/resources/bitsail-connector-unified-assert.json [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+  "name": "bitsail-connector-unified-assert",
```
Comment (garyli1019) [inline bitsail-connectors/connector-assert/src/main/resources/bitsail-connector-unified-assert.json:2]: should we remove `unified` here?

File: bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/executor/AssertRuleExecutor.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+import com.bytedance.bitsail.common.row.Row;
+import com.bytedance.bitsail.common.typeinfo.TypeInfo;
+import com.bytedance.bitsail.connector.assertion.sink.rule.AssertColumnRule;
+import com.bytedance.bitsail.connector.assertion.sink.rule.AssertSimpleRule;
+
+import java.util.List;
+import java.util.Map;
+
+public class AssertRuleExecutor {
```
Comment (garyli1019) [inline bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/executor/AssertRuleExecutor.java:27]: can we add UT for this class?

File: bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/AssertWriter.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  @Override
+  public void flush(boolean endOfInput) throws IOException {
+
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (rowRules != null) {
```
Comment (garyli1019) [inline bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/AssertWriter.java:70]: didn't get this. Why do we do this at closing

File: bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/rule/AssertRuleParser.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+import org.apache.commons.lang3.StringUtils;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class AssertRuleParser implements Serializable {
```
Comment (garyli1019) [inline bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/rule/AssertRuleParser.java:34]: IMO UT is necessary for this class. can we add it please

**Review by garyli1019 — COMMENTED** [review]
thanks for your contribution @liuxiaocs7 , left some comments

File: bitsail-connectors/connector-assert/src/main/resources/bitsail-connector-unified-assert.json (hunk 1 above)
Comment (liuxiaocs7) [reply bitsail-connectors/connector-assert/src/main/resources/bitsail-connector-unified-assert.json:2]: I noticed that connector V1 is uniformly named `bitsail-connector-unified-xxx`, like this, the same naming convention is used here.

image

File: bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/executor/AssertRuleExecutor.java (hunk 2 above)
Comment (liuxiaocs7) [reply bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/executor/AssertRuleExecutor.java:27]: sure, i'll add it

File: bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/rule/AssertRuleParser.java (hunk 4 above)
Comment (liuxiaocs7) [reply bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/rule/AssertRuleParser.java:34]: sure, by the way, what's the meaning of IMO? :)

File: bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/AssertWriter.java (hunk 3 above)
Comment (liuxiaocs7) [reply bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/AssertWriter.java:70]: `rowRules` is used to limit the number of processed data by engine, we need to ensure that the number of processed data is within the range of `[min_row, max_row]`. In `write` method, we add counter the after handler a row and compare it in `close` method finally.

Comment (garyli1019) [reply bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/AssertWriter.java:70]: so the row rule must be a single parallelism right, looks like this counter only count the number of rows in this writer.

File: bitsail-connectors/connector-assert/src/main/resources/bitsail-connector-unified-assert.json (hunk 1 above)
Comment (garyli1019) [reply bitsail-connectors/connector-assert/src/main/resources/bitsail-connector-unified-assert.json:2]: ok I see

File: bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/rule/AssertRuleParser.java (hunk 4 above)
Comment (garyli1019) [reply bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/rule/AssertRuleParser.java:34]: it stands for in my opinion :)

Comment (liuxiaocs7) [reply bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/rule/AssertRuleParser.java:34]: get it, thx

File: bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/AssertWriter.java (hunk 3 above)
Comment (liuxiaocs7) [reply bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/AssertWriter.java:70]: thanks for pointing this out, i think so, if we want to limit row number, the parallelism must be one, can we set it `writer_parallelism_num` by this conf, or any suggestions, thx

Comment (garyli1019) [reply bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/AssertWriter.java:70]: if the assert sink only support single parallelism, we should make sure the single parallelism was hard coded to ensure the correct result. The related class is `ParallelismComputable`. 
If we wanna support multi parallelism, I believe we should some how aggregate the count in the coordinator. We could make the multi parallelism support as a follow up task for future work.

Comment (liuxiaocs7) [reply bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/AssertWriter.java:70]: Thanks so much, got it, i'll look `ParallelismComputable` first.

**Comment by liuxiaocs7** [comment]
@garyli1019 The [ci failed](https://github.com/bytedance/bitsail/actions/runs/3799217233/jobs/6461536795), and error message shows failing to download maven dependency, which may be caused by network problems.

`Error:  Failed to execute goal on project bitsail-connector-rocketmq: Could not resolve dependencies for project com.bytedance.bitsail:bitsail-connector-rocketmq:jar:0.1.0-SNAPSHOT: Failed to collect dependencies at org.apache.rocketmq:rocketmq-client:jar:4.9.2: Failed to read artifact descriptor for org.apache.rocketmq:rocketmq-client:jar:4.9.2: Could not transfer artifact org.apache.rocketmq:rocketmq-client:pom:4.9.2 from/to central (https://repo.maven.apache.org/maven2): transfer failed for https://repo.maven.apache.org/maven2/org/apache/rocketmq/rocketmq-client/4.9.2/rocketmq-client-4.9.2.pom: Connection timed out (Read failed) -> [Help 1]`

**Review by garyli1019 — APPROVED** [review]
LGTM~ Thanks for your contribution @liuxiaocs7 , would you please make a follow up PR to add the documentation about this assert sink?

**Comment by liuxiaocs7** [comment]
> LGTM~ Thanks for your contribution @liuxiaocs7 , would you please make a follow up PR to add the documentation about this assert sink?

sure, I'd love to do

**Pull request merged**